controllers/serviceaccount_controller.go

```diff
@@ -249,74 +255,81 @@ func (r ServiceAccountReconciler) ReconcileNormal(guestClusterCtx *vmwarecontext
 		}
 	}()
 
-	pSvcAccounts, err := getProviderServiceAccounts(guestClusterCtx.ClusterContext)
+	pSvcAccounts, err := r.getProviderServiceAccounts(ctx, guestClusterCtx.ClusterContext)
 	if err != nil {
-		guestClusterCtx.Logger.Error(err, "Error fetching provider serviceaccounts")
+		log.Error(err, "Error fetching ProviderServiceAccounts")
 		return reconcile.Result{}, err
 	}
-	err = r.ensureProviderServiceAccounts(guestClusterCtx, pSvcAccounts)
+	err = r.ensureProviderServiceAccounts(ctx, guestClusterCtx, pSvcAccounts)
 	if err != nil {
-		guestClusterCtx.Logger.Error(err, "Error ensuring provider serviceaccounts")
+		log.Error(err, "Error ensuring ProviderServiceAccounts")
 		return reconcile.Result{}, err
 	}
 
 	return reconcile.Result{}, nil
 }
 
 // Ensure service accounts from provider spec is created.
-func (r ServiceAccountReconciler) ensureProviderServiceAccounts(guestClusterCtx *vmwarecontext.GuestClusterContext, pSvcAccounts []vmwarev1.ProviderServiceAccount) error {
+func (r *ServiceAccountReconciler) ensureProviderServiceAccounts(ctx context.Context, guestClusterCtx *vmwarecontext.GuestClusterContext, pSvcAccounts []vmwarev1.ProviderServiceAccount) error {
+	log := ctrl.LoggerFrom(ctx)
+
 	for i, pSvcAccount := range pSvcAccounts {
+		// Note: We have to use := here to not overwrite log & ctx outside the for loop.
+		log := log.WithValues("ProviderServiceAccount", klog.KRef(pSvcAccount.Namespace, pSvcAccount.Name))
+		ctx := ctrl.LoggerInto(ctx, log)
+
 		if guestClusterCtx.Cluster != nil && annotations.IsPaused(guestClusterCtx.Cluster, &(pSvcAccounts[i])) {
-			r.Logger.V(4).Info("ProviderServiceAccount %s/%s linked to a cluster that is paused or has pause annotation",
-				pSvcAccount.Namespace, pSvcAccount.Name)
+			log.V(4).Info("ProviderServiceAccount linked to a cluster that is paused or has pause annotation")
 			continue
 		}
 
 		// 1. Create service accounts by the name specified in Provider Spec
-		if err := r.ensureServiceAccount(guestClusterCtx.ClusterContext, pSvcAccount); err != nil {
+		if err := r.ensureServiceAccount(ctx, pSvcAccount); err != nil {
```

Comment:
We need a way to pass down the guestClusterCtx.Client or get a client from the cluster tracker

Reply:
ack! `guestClusterCtx.ClusterContext.Client` was the client to the MC, so using r.Client instead now which is 100% fine! Thx for claryfication